sycl/test/devicelib/assert.cpp

```diff
@@ -0,0 +1,200 @@
+// REQUIRES: cpu,linux
+// RUN: %clangxx -fsycl -c %s -o %t.o
+// RUN: %clangxx -fsycl %t.o %llvm_build_libs_dir/libsycl-glibc.o -o %t.out
+// (see the other RUN lines below; it is a bit complicated)
+//
+// assert() call in device code guarantees nothing: on some devices it behaves
+// in the usual way and terminates the program. On other devices it can print an
+// error message and *continue* execution. Less capable devices can even ignore
+// an assert!
+//
+// This makes testing an assert() a bit difficult task, and we have to rely on
+// the implementation details to make sure that both "native" and "fallback"
+// implementations work as expected.
+//
+// This test works only on Intel OpenCL CPU implementation, which is known to
+// behave as follows:
+//
+//   Fallback mode (aka the best we can do by following the OpenCL spec):
+//     1. Assertion condition is printed to *stdout* by the OpenCL printf().
+//     2. Process (both host and device) is terminated by a SIGSEGV.
```

Comment:
I am lost here. The fallback in SYCL is just the normal C++ code on the host so I have the feeling it should be the same for a "normal" `assert()` of the system.
What is the expectation of the end-user? That some functions are just changed by the Intel implementation when in kernels on the host? How a programmer can be sure that we can call the right function?

Reply:
> The fallback in SYCL is just the normal C++ code on the host so I have the feeling it should be the same for a "normal" `assert()` of the system.

The fallback in this context is a generic implementation of standard library functions provided in a SPIR-V form. This implementation is not aware about device-specific features, and can only use what is available in SPIR-V (i.e. no termination is possible for assert).

If a device compiler can do something better than a generic implementation, it can override some functions to achieve performance or functional improvements.

Anyway, the line 20 is misleading (2. Process is terminated). The SIGSEGV is not a side-effect from the fallback implementation: an `unreachable` instruction follows an `__assert_fail` call, so the process terminates after reaching it.